NEP: Make like= argument added in NEP-35 strict

pentschev opened the pull request:
This PR makes the `like=` argument in array creation functions strict, meaning that dispatching on a reference object that does _NOT_ implement the `__array_function__` protocol will raise an exception, as per discussion in https://github.com/numpy/numpy/issues/17075 .

I'm not sure if by only allowing objects implementing the `__array_function__` here is excessively strict and we should instead consider having an exclusion list (e.g., `list`s, `tuple`s, etc.) or if this is the appropriate approach here.

cc @seberg @rgommers @hameerabbasi @shoyer @quasiben @jakirkham

pentschev committed:
MAINT: Add array_function_dispatch_like helper

This helper is used to dispatch a Python function with the like=
argument or raise an error if the object passed doesn't implement the
__array_function__ protocol.

pentschev committed:
MAINT: Make like= in Python functions strict

Only allow objects that implement __array_function__

pentschev committed:
MAINT: Make like= in np.array() strict

Only allow objects that implement __array_function__

pentschev committed:
TST: Test strict like=

BvB93 commented on numpy/core/overrides.py:
```diff
@@ -34,6 +34,15 @@ def set_array_function_like_doc(public_api):
 
+def array_function_dispatch_like(func, *args, **kwargs):
+    if not hasattr(kwargs['like'], '__array_function__'):
+        raise ValueError(
```
Shouldn't this be a `TypeError`?
Considering only the object's type is relevant here; not its value.

eric-wieser replied:
On that note, this should probably be `if not hasattr(type(wargs['like']), '__array_function__'):`

pentschev replied:
> Shouldn't this be a `TypeError`?
> Considering only the object's type is relevant here; not its value.

I wasn't really sure which of those would be a better fit, but I agree `TypeError` is a good choice, I'll change that.

pentschev replied:
> On that note, this should probably be `if not hasattr(type(kwargs['like']), '__array_function__'):`

Why? `__array_function__` is a method of the instance too. I believe the suggested change will also work, but I don't really see the benefit or reason why that's necessary.

eric-wieser replied:
> Why? `__array_function__` is a method of the instance too.

Yes, but an object that has  `__array_function__` on the instance but not on the type will not actually go through the dispatch procedure.

pentschev replied:
Got it, I didn't realize that's what `PyArray_LookupSpecial` does. These two changes are in now.

eric-wieser commented on numpy/core/src/multiarray/arrayfunction_override.c:
```diff
@@ -379,11 +379,16 @@ array_implement_c_array_function_creation(
+    if (like_arg == NULL) {
         return NULL;
     }
+    else if (!PyObject_HasAttrString(like_arg, "__array_function__")) {
```
This will lead to weird situations where this check passes, but `__array_function__` is not actually present - to behave accurately, this should use `get_array_function(like_arg)`

pentschev replied:
Good idea, I'll change that.

pentschev replied:
This change is in too.

pentschev committed:
MAINT: Change like= exception to TypeError

pentschev committed:
MAINT: Use get_array_function with like=

pentschev committed:
MAINT: Check like= type for __array_function__

seberg commented on numpy/core/src/multiarray/arrayfunction_override.c:
```diff
@@ -379,11 +379,16 @@ array_implement_c_array_function_creation(
         return NULL;
     }
+    else if (!get_array_function(like_arg)) {
+        return PyErr_Format(
+            PyExc_TypeError,
```
```suggestion
        return PyErr_Format(PyExc_TypeError,
```
We usually put the exception on the same line.  (I never realize that these function return NULL, but that looks nice).

seberg reviewed (commented):
Sorry for not looking at this earlier.  Looks generally good, but on the C-side in `array_implement_array_function` we already have the equivalence of:
```
if "like" in kwargs:
     del kwargs["like"]
```
so I think we can do the check just as well there; Or am I missing something?

A simple test (at least one python function and one C-side one) would be nice as well.

pentschev committed:
STY: Move exception to same line of PyErr_Format

Co-authored-by: Sebastian Berg <sebastian@sipsolutions.net>

pentschev committed:
REV: Make like= in Python functions strict

pentschev committed:
REV: Add array_function_dispatch_like helper

pentschev committed:
MAINT: Make like= strict in array_implement_array_function

If like= is passed but doesn't implement __array_function__, raise a
TypeError.

pentschev commented:
> Sorry for not looking at this earlier. Looks generally good, but on the C-side in `array_implement_array_function` we already have the equivalence of:
> 
> ```
> if "like" in kwargs:
>      del kwargs["like"]
> ```
> 
> so I think we can do the check just as well there; Or am I missing something?

That's a good point, I've reverted the original changes and moved that to `array_implement_array_function`.

> A simple test (at least one python function and one C-side one) would be nice as well.

I don't think I understand this comment, `test_no_array_function_like` will test all Python (e.g., `np.asarray`, etc.) and C (e.g., `np.array`, etc.) functions for a strict `like=` parameter. The test existed before but it wasn't strict, I now repurposed it to check that `TypeError` is raised in all cases. So I think what you ask for is already covered, no?

seberg commented:
Thanks, and sorry about the test confusion, my silly oversight.

hameerabbasi commented on numpy/core/src/multiarray/arrayfunction_override.c:
```diff
@@ -342,9 +342,15 @@ array_implement_array_function(
+        PyObject *like_arg = PyDict_GetItem(kwargs, npy_ma_str_like);
+        if (like_arg && !get_array_function(like_arg)) {
+            return PyErr_Format(PyExc_TypeError,
+                "The `like` object must implement the `__array_function__` protocol");
```
```suggestion
                "The `type(like)` must implement the `__array_function__` protocol");
```

seberg replied:
Good point to look at the message closer. Not sure like adding `type` it seems like unnecessary precision. I would suggest using the same jargon words we use in the documentation:

> must be an array-like that implements the ``__array_function__`` protocol, but is an instance of %S

(passing `Py_TYPE(obj)` for the format code. Although types print as `<class 'name'>` which is slightly annoying, and printing just the name was a bit more involved. So I am fine also with just not printing at all.)

hameerabbasi replied:
I'd say "expected" instead of "must be", but the rest is better!

hameerabbasi commented on numpy/core/tests/test_overrides.py:
```diff
@@ -527,33 +527,21 @@ def test_array_like(self, function, args, kwargs, numpy_ref):
-                array_like.fill(1)
 
-            assert_equal(array_like, np_arr)
+        with assert_raises(TypeError):
```
I'd add a string match here to ensure it's the right `TypeError`.

hameerabbasi reviewed (approved):
LGTM, just two minor changes.

pentschev commented:
Thanks @hameerabbasi for the suggestions, both are in now!

seberg commented:
Not sure its a real improvement, but couldn't help it and tweaked the message. Will merge when tests succeeds, if anyone is unhappy with it, please kick me and I will follow up :).

pentschev commented:
> Not sure its a real improvement, but couldn't help it and tweaked the message. Will merge when tests succeeds, if anyone is unhappy with it, please kick me and I will follow up :).

I'm fine with either one of them, thanks @seberg !

pentschev commented:
Thanks everyone here for reviews and @seberg for merging!
